src/PathMappingRepository.php

```diff
@@ -0,0 +1,547 @@
+    {
+        if ('glob' !== $language) {
+            throw UnsupportedLanguageException::forLanguage($language);
+        }
+
+        Assert::stringNotEmpty($query, 'The glob must be a non-empty string. Got: %s');
+        Assert::startsWith($query, '/', 'The glob %s is not absolute.');
+
+        $query = Path::canonicalize($query);
+
+        return $this->search($query, true)->count() > 0;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function add($path, $resource)
+    {
+        Assert::stringNotEmpty($path, 'The path must be a non-empty string. Got: %s');
+        Assert::startsWith($path, '/', 'The path %s is not absolute.');
+
+        $path = Path::canonicalize($path);
+
+        if ($resource instanceof ResourceCollection) {
+            $this->ensureDirectoryExists($path);
+
+            foreach ($resource as $child) {
+                $this->addResource($path . '/' . $child->getName(), $child);
+            }
+
+            $this->sortStore();
+
+            return;
+        }
+
+        if ($resource instanceof FilesystemResource) {
+            $this->ensureDirectoryExists(Path::getDirectory($path));
+            $this->addResource($path, $resource);
+            $this->sortStore();
+
+            return;
+        }
+
+        throw new UnsupportedResourceException(sprintf(
+            'The passed resource must be a FilesystemResource or a ResourceCollection. Got: %s',
+            is_object($resource) ? get_class($resource) : gettype($resource)
+        ));
+    }
+
+    /**
+     * Removes all resources matching the given query.
+     *
+     * This method is not supported by this repository.
+     *
+     * @param string $query    A resource query.
+     * @param string $language The language of the query. All implementations
+     *                         must support the language "glob".
+     *
+     * @return integer The number of resources removed from the repository.
+     *
+     * @throws BadMethodCallException
+     */
+    public function remove($query, $language = 'glob')
+    {
+        throw new BadMethodCallException(sprintf('%s\(\) is not supported in %s', __METHOD__, __CLASS__));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clear()
+    {
+        // Subtract root
+        $removed = $this->countStore() - 1;
+
+        $this->store->clear();
+        $this->createRoot();
+
+        return $removed;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listChildren($path)
+    {
+        return new ArrayResourceCollection($this->getChildren($this->get($path)));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasChildren($path)
+    {
+        return $this->getChildren($this->get($path))->count() !== 0;
+    }
+
+
+    /**
+     * Find a resource by its path.
+     *
+     * @param string $path The path to resolve.
+     *
+     * @return \Puli\Repository\Api\Resource\Resource|null  The resource or null if
```

Comment:
Apart from that, Resource may have to be renamed anyway if it really becomes a reserved keyword in PHP7 :P